### PR TITLE
Authorization

### LINKED ISSUE #14: Authorization
#### As an application hoster I want to restrict some functionality to specific users.

**Story**:

As discussed previously, we want to prevent malicious users from destroying
content they did not add to our site. We have already built the foundation for
registering and authenticating users in our system. Now it is time to leverage
that authentication and provide restrictions based on the state of the
user.

The two primary states a user can belong to are unauthenticated, and
authenticated. With regard to individual images, the two substates an
authenticated user can belong to are owner or viewer, depending on whether or
not they added the image.

**Acceptance criteria**:
- [x] Logged out users are restricted to read-only access.
- [x] Links to non-read actions are not shown to logged out users.
- [x] Attempting to access a page that requires authentication redirects you to
  the login page.
- [x] Upon successful login, you are redirected back to where you had tried to
  navigate.
- [x] Every image is associated with the user that added it.
- [x] Only the user that added an image can delete that image.

**Discussion topics**:
- [ ] If we wanted to add a lot of authorization rules, how might we want to
  organize our code to keep it maintainable?

**Further reading**:
- https://github.com/elabs/pundit

**Dependencies**:
- Authentication

### PR DESCRIPTION
**Pull request by ragurney**
Closes #14

### CODE REVIEW

File: test/integration/image_policy_integration_test.rb [hunk 1]
```diff
@@ -61,7 +61,54 @@ def setup
     assert_select 'a', text: 'New Image', count: 2
   end
 
-  test 'new/create :previous_path param should still be present after invalid session submitted error' do
+  test 'delete authorization error -> login -> successfully delete image' do
```
Comment (jyoder) [inline test/integration/image_policy_integration_test.rb]: The name of this test is confusing to me. It looks like you never successfully delete the image (based on the assertions), but the name of the test indicates that the image is successfully deleted.

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -31,15 +33,31 @@ def session_params
...
+  def handle_user_login(user)
+    log_in(user)
+    remember(user) if @session.remember_me == '1'
+  end
+
+  def successful_login_redirect_path
+    params[:previous_path] || root_path
+  end
+
+  def set_flash_if_previous_action_not_get
```
Comment (jyoder) [inline app/controllers/sessions_controller.rb]: I'm a little confused about the purpose of this. Maybe we can chat real quick and you can explain it to me.

File: app/views/sessions/new.html.erb [hunk 3]
```diff
@@ -10,6 +10,10 @@
         <%= f.input :password, label: 'Enter password', placeholder: 'password'%>
         <%= f.input :remember_me, as: :boolean, inline_label: 'Remember me' %>
         <%= f.button :submit, 'Log in', class: 'btn btn-primary' %>
+
+        <%= hidden_field_tag :previous_path, @previous_path || params[:previous_path] %>
```
Comment (jyoder) [inline app/views/sessions/new.html.erb]: Again, a bit confused about this. Explain to me in person.

File: app/views/images/new.html.erb [hunk 4]
```diff
@@ -4,7 +4,7 @@
   <div class="col-xs-6">
     <%= simple_form_for @image do |f| %>
       <%= f.error_notification  %>
-
+      <%= f.input :user_id, as: :hidden, input_html: { value: current_user.id } %>
```
Comment (ragurney) [inline app/views/images/new.html.erb]: bdd5f370571c976ccfd640451ad5f0b7b21e1af4

File: app/controllers/sessions_controller.rb (hunk 2 above)
Comment (ragurney) [inline app/controllers/sessions_controller.rb]: f96bc1b, 4c73857, a9a5436

File: app/views/sessions/new.html.erb (hunk 3 above)
Comment (ragurney) [inline app/views/sessions/new.html.erb]: f96bc1b, 4c73857, a9a5436

**Comment by ragurney** [comment]
Pushed some commits that incorporate your suggestions.

**Comment by jyoder** [comment]
Your changes look good to me! 👍

**Pull request merged**